src/Microsoft.AspNet.OData.Shared/Query/SkipTokenQueryOption.cs

```diff
@@ -0,0 +1,168 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.AspNet.OData.Query.Validators;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OData;
+using Microsoft.OData.Edm;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    /// <summary>
+    /// This defines a $skiptoken OData query option for querying.
+    /// </summary>
+    public class SkipTokenQueryOption
+    {
+        private string _value;
+        private ODataQueryOptionParser _queryOptionParser;
+
+        /// <summary>
+        /// Initialize a new instance of <see cref="SkipQueryOption"/> based on the raw $skiptoken value and
+        /// an EdmModel from <see cref="ODataQueryContext"/>.
+        /// </summary>
+        /// <param name="rawValue">The raw value for $skiptoken query. It can be null or empty.</param>
+        /// <param name="context">The <see cref="ODataQueryContext"/> which contains the <see cref="IEdmModel"/> and some type information</param>
+        /// <param name="queryOptionParser">The <see cref="ODataQueryOptionParser"/> which is used to parse the query option.</param>
+        public SkipTokenQueryOption(string rawValue, ODataQueryContext context, ODataQueryOptionParser queryOptionParser)
+        {
+            if (context == null)
+            {
+                throw Error.ArgumentNull("context");
+            }
+
+            if (String.IsNullOrEmpty(rawValue))
+            {
+                throw Error.ArgumentNullOrEmpty("rawValue");
+            }
+
+            if (queryOptionParser == null)
+            {
+                throw Error.ArgumentNull("queryOptionParser");
+            }
+
+            RawValue = rawValue;
+            Validator = context.GetSkipTokenQueryValidator();
+            SkipTokenHandler = context.GetSkipTokenHandler();
+            Context = context;
+            _queryOptionParser = queryOptionParser;
+        }
+
+        // This constructor is intended for unit testing only.
+        internal SkipTokenQueryOption(string rawValue, ODataQueryContext context)
+        {
+            if (context == null)
+            {
+                throw Error.ArgumentNull("context");
+            }
+
+            if (String.IsNullOrEmpty(rawValue))
+            {
+                throw Error.ArgumentNullOrEmpty("rawValue");
+            }
+
+            RawValue = rawValue;
+            Validator = context.GetSkipTokenQueryValidator();
+            _queryOptionParser = new ODataQueryOptionParser(
+                context.Model,
+                context.ElementType,
+                context.NavigationSource,
+                new Dictionary<string, string> { { "$skiptoken", rawValue } },
+                context.RequestContainer);
+        }
```

Comment:
If this is just for unit tests, can we remove this and create the ODataQueryOptionParser in the unit test (if/when added).